include/alpaka/extent/Traits.hpp

```diff
@@ -22,121 +23,183 @@ namespace alpaka
-    {
-        setExtent<Dim<TExtent>::value - 2u>(extent, height);
-    }
-    //! Sets the depth.
-    ALPAKA_NO_HOST_ACC_WARNING
-    template<typename TExtent, typename TDepth>
-    ALPAKA_FN_HOST_ACC auto setDepth(TExtent& extent, TDepth const& depth) -> void
-    {
-        setExtent<Dim<TExtent>::value - 3u>(extent, depth);
-    }
+        ALPAKA_NO_HOST_ACC_WARNING
+        ALPAKA_FN_HOST_ACC static constexpr auto setExtent(Vec<TDim, TVal>& extent, TExtentVal const& extentVal)
+            -> void
+        {
+            extent[TIdxIntegralConst::value] = extentVal;
+        }
+    };
+} // namespace trait
 
-    // Trait specializations for unsigned integral types.
-    namespace trait
+//! Sets the extent in the given dimension.
+ALPAKA_NO_HOST_ACC_WARNING
+template<std::size_t Tidx, typename TExtent, typename TExtentVal>
+ALPAKA_FN_HOST_ACC auto setExtent(TExtent& extent, TExtentVal const& extentVal) -> void
+{
+    trait::SetExtent<DimInt<Tidx>, TExtent, TExtentVal>::setExtent(extent, extentVal);
+}
+//! Sets the width.
+ALPAKA_NO_HOST_ACC_WARNING
+template<typename TExtent, typename TWidth>
+ALPAKA_FN_HOST_ACC auto setWidth(TExtent& extent, TWidth const& width) -> void
+{
+    setExtent<Dim<TExtent>::value - 1u>(extent, width);
+}
+//! Sets the height.
+ALPAKA_NO_HOST_ACC_WARNING
+template<typename TExtent, typename THeight>
+ALPAKA_FN_HOST_ACC auto setHeight(TExtent& extent, THeight const& height) -> void
+{
+    setExtent<Dim<TExtent>::value - 2u>(extent, height);
+}
+//! Sets the depth.
+ALPAKA_NO_HOST_ACC_WARNING
+template<typename TExtent, typename TDepth>
+ALPAKA_FN_HOST_ACC auto setDepth(TExtent& extent, TDepth const& depth) -> void
+{
+    setExtent<Dim<TExtent>::value - 3u>(extent, depth);
+}
+
+// Trait specializations for integral types.
+namespace trait
+{
+    template<typename Integral>
+    struct GetExtents<Integral, std::enable_if_t<std::is_integral_v<Integral>>>
     {
-        //! The unsigned integral width get trait specialization.
-        template<typename TExtent>
-        struct GetExtent<DimInt<0u>, TExtent, std::enable_if_t<std::is_integral_v<TExtent>>>
+        ALPAKA_NO_HOST_ACC_WARNING
+        ALPAKA_FN_HOST_ACC auto operator()(Integral i) const
         {
-            ALPAKA_NO_HOST_ACC_WARNING
-            ALPAKA_FN_HOST_ACC static auto getExtent(TExtent const& extent) -> Idx<TExtent>
-            {
-                return extent;
-            }
-        };
-        //! The unsigned integral width set trait specialization.
-        template<typename TExtent, typename TExtentVal>
-        struct SetExtent<DimInt<0u>, TExtent, TExtentVal, std::enable_if_t<std::is_integral_v<TExtent>>>
+            return Vec{i};
+        }
+    };
+    //! The unsigned integral width set trait specialization.
+    template<typename TExtent, typename TExtentVal>
+    struct SetExtent<DimInt<0u>, TExtent, TExtentVal, std::enable_if_t<std::is_integral_v<TExtent>>> // WTF?
```

Comment:
Why the `WTF` here?

Reply:
Because I question this specialization in its entirety. Good you saw that, let's have a discussion.

What is the extent of an `int`? Is it `0`? Is it `1`? What if the `int` holds the value `42`, is `42` the extent of the `int` then?

Reply:
> What is the extent of an `int`? Is it `0`?

yes

Reply:
> > What is the extent of an `int`? Is it `0`?
> 
> yes

Well, too bad! alpaka defines it (since at least 2015) to be the int's value :)

Reply:
I had a quick look if we can take the specializations for integrals away, but they are way too deeply used inside alpaka. workdivs, buffer allocation, etc. All that uses the extents API to work with integer sizes etc.

Reply:
Okay, that is what I was afraid of - some arcane usage. So we are also relying on this functionality of returning the integer value?

Reply:
I removed the `WTF` comment, and will leave extents for integers as they are.

Reply:
@j-stephan totally! `alpaka::allocBuf(.., 42)` uses that to turn `42` into a `Vec` internally. You could refactor this but you would have to touch a few places.